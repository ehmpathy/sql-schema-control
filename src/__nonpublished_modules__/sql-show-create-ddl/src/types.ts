export interface DatabaseConnection {
  query: (args: { sql: string; values?: (string | number)[] }) => Promise<{ rows: any[] }>;
  end: () => Promise<void>;
}
