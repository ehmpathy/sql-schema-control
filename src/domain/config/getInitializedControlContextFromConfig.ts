/*
  1. get the config (validated and hydrated)
  2. initialize the environment based on the config
    - start the db connection
    - initialize the control environment
  3. return the language+dialect, connection, and definitions
*/
