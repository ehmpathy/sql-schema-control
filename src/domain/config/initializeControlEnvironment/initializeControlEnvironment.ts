/*
  complete any actions required to initialize control environment
    1. connect to db
    2. provision change log table
*/
