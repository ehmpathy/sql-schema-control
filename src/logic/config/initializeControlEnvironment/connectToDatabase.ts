import mysql from 'mysql2/promise';
import { DatabaseConnection, DatabaseLanguage, ControlConfig, ConnectionConfig } from '../../../types';

// create the connection to database
const connectionAdapters = {
  [DatabaseLanguage.MYSQL]: async ({ connectionConfig }: { connectionConfig: ConnectionConfig }): Promise<DatabaseConnection> => {
    const connection = await mysql.createConnection({
      host: connectionConfig.host,
      port: connectionConfig.port,
      database: connectionConfig.schema,
      user: connectionConfig.username,
      password: connectionConfig.password,
    });
    return {
      query: async ({ sql, values }: { sql: string, values?: any[] }) => {
        return connection.query(sql, values);
      },
      end: async () => connection.end(),
    };
  },
};

export const connectToDatabase = async ({ config }: { config: ControlConfig }): Promise<DatabaseConnection> => {
  // 1. get the connection adapter for the method
  const getDbConnection = connectionAdapters[config.language];

  // 2. attempt to connect
  const connection = await getDbConnection({ connectionConfig: config.connection });

  // 3. run a test query to check connection
  await connection.query({ sql: 'SHOW TABLES' }); // TODO: make this not mysql specific

  // 4. return the connection
  return connection;
};
