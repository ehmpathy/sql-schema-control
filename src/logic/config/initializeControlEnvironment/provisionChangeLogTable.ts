import { DatabaseConnection } from '../../../types';
import { readFileAsync } from '../_utils/readFileAsync';

export const provisionChangeLogTable = async ({ connection }: { connection: DatabaseConnection }) => {
  // create the table if not exists (sql consideres its existance for us)
  const createTableSql = await readFileAsync({
    filePath: `${__dirname}/../../../../schema/tables/schema_control_change_log.mysql.sql`,
  }); // TODO: make independent of language (mysql currently)
  await connection.query({ sql: createTableSql });
};
