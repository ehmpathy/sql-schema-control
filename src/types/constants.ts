export enum DefinitionType {
  RESOURCE = 'resource',
  CHANGE = 'change',
}

export enum DatabaseLanguage {
  MYSQL = 'mysql',
}

export interface DatabaseConnection {
  query: (args: { sql: string, values?: any[] }) => Promise<any>;
  end: () => Promise<void>;
}

export enum ChangeDefinitionStatus {
  NOT_APPLIED = 'NOT_APPLIED',
  UP_TO_DATE = 'UP_TO_DATE',
  OUT_OF_DATE = 'OUT_OF_DATE',
}
