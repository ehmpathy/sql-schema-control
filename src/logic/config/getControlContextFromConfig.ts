import { ControlContext, ResourceDefinition } from '../../types';
import { getUncontrolledResources } from '../schema/resourceDefinition/getUncontrolledResources';
import { getConfig } from './getConfig';
import { initializeControlEnvironment } from './initializeControlEnvironment';

/*
  1. get the config (validated and hydrated)
  2. initialize the environment based on the config
    - start the db connection
    - initialize the control environment
  3. return the control context
*/
export const getControlContextFromConfig = async ({
  configPath,
  strict = false,
}: {
  configPath: string;
  strict?: boolean;
}) => {
  // 1. get the config
  const config = await getConfig({ configPath });

  // 2. initialize the environment based on the config
  const { connection } = await initializeControlEnvironment({ config });

  // 3. define the definitions such that we can append to them if we want (e.g., if want to add uncontrolled)
  const definitions = [...config.definitions];

  // 4. get uncontrolled resources, if strict
  if (config.strict || strict) {
    const controlledResources = definitions.filter(
      (def) => def.constructor === ResourceDefinition,
    ) as ResourceDefinition[];
    const uncontrolledResources = await getUncontrolledResources({ connection, controlledResources });
    definitions.push(...uncontrolledResources); // append the uncontrolled resources to the definitions
  }

  // 5. return the control context
  return new ControlContext({
    language: config.language,
    dialect: config.dialect,
    connection, // note: the connection is a live dbConnection
    definitions, // note: uncontrolled resources will already have statuses
  });
};
