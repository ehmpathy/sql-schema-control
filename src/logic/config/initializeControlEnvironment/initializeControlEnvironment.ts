import { ControlConfig } from '../../../domain';
import { connectToDatabase } from './connectToDatabase';
import { provisionChangeLogTable } from './provisionChangeLogTable';

/*
  complete any actions required to initialize control environment
    1. connect to db
    2. provision change log table
*/
export const initializeControlEnvironment = async ({
  config,
}: {
  config: ControlConfig;
}) => {
  // 1. connect to db
  const connection = await connectToDatabase({ config });

  // 2. provision the change log table
  await provisionChangeLogTable({ connection });

  // 3. return the connection
  return { connection };
};
