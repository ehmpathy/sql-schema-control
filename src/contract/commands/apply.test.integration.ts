import Apply from './apply';
import { stdout } from 'stdout-stderr';
import { DatabaseConnection, DatabaseLanguage, ControlConfig } from '../../types';
import { promiseConfig } from '../_test_assets/connection.config';
import { initializeControlEnvironment } from '../../logic/config/initializeControlEnvironment';

describe('apply', () => {
  let connection: DatabaseConnection;
  beforeAll(async () => {
    const config = new ControlConfig({
      language: DatabaseLanguage.MYSQL,
      dialect: '5.7',
      connection: await promiseConfig(),
      definitions: [],
      strict: false,
    });
    ({ connection } = await initializeControlEnvironment({ config })); // ensure db is provisioned and get connection
  });
  afterAll(async () => {
    await connection.end();
  });
  it('should have an expected appearance when all changes need to be applied', async () => {
    // ensure previous runs dont break this test
    await connection.query({ sql: 'DROP USER IF EXISTS user_name' });
    await connection.query({ sql: 'DROP TABLE IF EXISTS data_source' });
    await connection.query({ sql: 'DROP TABLE IF EXISTS notification_version' });
    await connection.query({ sql: 'DROP TABLE IF EXISTS notification' });
    await connection.query({ sql: 'DROP FUNCTION IF EXISTS find_message_hash_by_text' });
    await connection.query({ sql: 'DROP PROCEDURE IF EXISTS upsert_message' });
    await connection.query({ sql: 'DELETE FROM schema_control_change_log' });

    // run the test
    process.stdout.isTTY = undefined; // since listr acts differently if nonTTY and jest is nonTTY when more than one test is run
    stdout.stripColor = false; // dont strip color
    // stdout.print = true;
    stdout.start();
    await Apply.run(['-c', `${__dirname}/../_test_assets/control.yml`]);
    stdout.stop();
    const output = stdout.output
      .split('\n').filter(line => !line.includes('console.log')).join('\n') // strip the console log portion
      .replace(/\[\d\d:\d\d:\d\d\]/g, ''); // remove all timestamps, since they change over time...
    expect(output).toMatchSnapshot();
    process.stdout.isTTY = true;
  });
  it('should have an expected appearance when all changes need to be reapplied, if possible', async () => {
    // ensure previous runs dont break this test
    await connection.query({ sql: 'DROP USER IF EXISTS user_name' });
    await connection.query({ sql: 'DROP TABLE IF EXISTS data_source' });
    await connection.query({ sql: 'DROP TABLE IF EXISTS notification_version' });
    await connection.query({ sql: 'DROP TABLE IF EXISTS notification' });
    await connection.query({ sql: 'DROP FUNCTION IF EXISTS find_message_hash_by_text' });
    await connection.query({ sql: 'DROP PROCEDURE IF EXISTS upsert_message' });
    await connection.query({ sql: 'DELETE FROM schema_control_change_log' });

    // apply the definitions the first time
    await Apply.run(['-c', `${__dirname}/../_test_assets/control.yml`]);

    // reapply the definitions
    process.stdout.isTTY = undefined; // since listr acts differently if nonTTY and jest is nonTTY when more than one test is run
    stdout.stripColor = false; // dont strip color
    // stdout.print = true;
    stdout.start();
    await Apply.run(['-c', `${__dirname}/../_test_assets/control.yml`]);
    stdout.stop();
    const output = stdout.output
      .split('\n').filter(line => !line.includes('console.log')).join('\n') // strip the console log portion
      .replace(/\[\d\d:\d\d:\d\d\]/g, ''); // remove all timestamps, since they change over time...
    expect(output).toMatchSnapshot();
    process.stdout.isTTY = true;
  });
});
