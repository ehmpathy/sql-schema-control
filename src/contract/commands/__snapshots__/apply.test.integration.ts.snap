// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apply mysql should have an expected appearance when all changes need to be applied 1`] = `
"    [1mApplying required actions...[22m

    

      [1m[32mâœ”[39m[22m [1m[1m[32m[APPLY][39m[1m src/contract/__test_assets__/mysql/definitions/tables/data_source.sql [90m(change:table_20190626_1)[39m[22m

      [1m[32mâœ”[39m[22m [1m[1m[32m[APPLY][39m[1m src/contract/__test_assets__/mysql/definitions/tables/notification.sql [90m(resource:table:notification)[39m[22m

      [1m[32mâœ”[39m[22m [1m[1m[32m[APPLY][39m[1m src/contract/__test_assets__/mysql/definitions/tables/notification_version.sql [90m(resource:table:notification_version)[39m[22m

      [1m[32mâœ”[39m[22m [1m[1m[32m[APPLY][39m[1m src/contract/__test_assets__/mysql/definitions/tables/spaceship.sql [90m(resource:table:spaceship)[39m[22m

      [1m[32mâœ”[39m[22m [1m[1m[32m[APPLY][39m[1m src/contract/__test_assets__/mysql/definitions/tables/spaceship_cargo.sql [90m(resource:table:spaceship_cargo)[39m[22m

      [1m[32mâœ”[39m[22m [1m[1m[32m[APPLY][39m[1m src/contract/__test_assets__/mysql/definitions/views/view_spaceship_with_cargo.sql [90m(resource:view:view_spaceship_with_cargo)[39m[22m

      [1m[32mâœ”[39m[22m [1m[1m[32m[APPLY][39m[1m src/contract/__test_assets__/mysql/definitions/functions/find_message_hash_by_text.sql [90m(resource:function:find_message_hash_by_text)[39m[22m

      [1m[32mâœ”[39m[22m [1m[1m[32m[APPLY][39m[1m src/contract/__test_assets__/mysql/definitions/procedures/upsert_message.sql [90m(resource:procedure:upsert_message)[39m[22m

      [1m[32mâœ”[39m[22m [1m[1m[32m[APPLY][39m[1m src/contract/__test_assets__/mysql/definitions/init/service_user.sql [90m(change:init_service_user)[39m[22m

      [1m[32mâœ”[39m[22m [1m[1m[32m[APPLY][39m[1m src/contract/__test_assets__/mysql/definitions/init/data_sources.sql [90m(change:init_data_source)[39m[22m

"
`;

exports[`apply mysql should have an expected appearance when all changes need to be reapplied, if possible 1`] = `
"    [1mApplying required actions...[22m

    

      [1m[33mâ†“[39m[22m [1m[1m[31m[MANUAL_MIGRATION][39m[1m src/contract/__test_assets__/mysql/definitions/tables/notification.sql [90m(resource:table:notification)[39m[22m

      [1m[33mâ†“[39m[22m [1m[1m[31m[MANUAL_MIGRATION][39m[1m src/contract/__test_assets__/mysql/definitions/tables/notification_version.sql [90m(resource:table:notification_version)[39m[22m

      [1m[33mâ†“[39m[22m [1m[1m[31m[MANUAL_MIGRATION][39m[1m src/contract/__test_assets__/mysql/definitions/tables/spaceship.sql [90m(resource:table:spaceship)[39m[22m

      [1m[33mâ†“[39m[22m [1m[1m[31m[MANUAL_MIGRATION][39m[1m src/contract/__test_assets__/mysql/definitions/tables/spaceship_cargo.sql [90m(resource:table:spaceship_cargo)[39m[22m

      [1m[32mâœ”[39m[22m [1m[1m[33m[REAPPLY][39m[1m src/contract/__test_assets__/mysql/definitions/views/view_spaceship_with_cargo.sql [90m(resource:view:view_spaceship_with_cargo)[39m[22m

      [1m[32mâœ”[39m[22m [1m[1m[33m[REAPPLY][39m[1m src/contract/__test_assets__/mysql/definitions/functions/find_message_hash_by_text.sql [90m(resource:function:find_message_hash_by_text)[39m[22m

      [1m[32mâœ”[39m[22m [1m[1m[33m[REAPPLY][39m[1m src/contract/__test_assets__/mysql/definitions/procedures/upsert_message.sql [90m(resource:procedure:upsert_message)[39m[22m

"
`;

exports[`apply postgres should have an expected appearance when all changes need to be applied 1`] = `
"    [1mApplying required actions...[22m

    

      [1m[32mâœ”[39m[22m [1m[1m[32m[APPLY][39m[1m src/contract/__test_assets__/postgres/definitions/tables/data_source.sql [90m(change:table_20190626_1)[39m[22m

      [1m[32mâœ”[39m[22m [1m[1m[32m[APPLY][39m[1m src/contract/__test_assets__/postgres/definitions/tables/photo.sql [90m(resource:table:photo)[39m[22m

      [1m[32mâœ”[39m[22m [1m[1m[32m[APPLY][39m[1m src/contract/__test_assets__/postgres/definitions/tables/spaceship.sql [90m(resource:table:spaceship)[39m[22m

      [1m[32mâœ”[39m[22m [1m[1m[32m[APPLY][39m[1m src/contract/__test_assets__/postgres/definitions/tables/spaceship_cargo.sql [90m(resource:table:spaceship_cargo)[39m[22m

      [1m[32mâœ”[39m[22m [1m[1m[32m[APPLY][39m[1m src/contract/__test_assets__/postgres/definitions/views/view_spaceship_with_cargo.sql [90m(resource:view:view_spaceship_with_cargo)[39m[22m

      [1m[32mâœ”[39m[22m [1m[1m[32m[APPLY][39m[1m src/contract/__test_assets__/postgres/definitions/functions/upsert_photo.sql [90m(resource:function:upsert_photo)[39m[22m

      [1m[32mâœ”[39m[22m [1m[1m[32m[APPLY][39m[1m src/contract/__test_assets__/postgres/definitions/init/service_user.sql [90m(change:init_service_user)[39m[22m

      [1m[32mâœ”[39m[22m [1m[1m[32m[APPLY][39m[1m src/contract/__test_assets__/postgres/definitions/init/data_sources.sql [90m(change:init_data_source)[39m[22m

"
`;

exports[`apply postgres should have an expected appearance when all changes need to be reapplied, if possible 1`] = `
"    [1mApplying required actions...[22m

    

      [1m[33mâ†“[39m[22m [1m[1m[31m[MANUAL_MIGRATION][39m[1m src/contract/__test_assets__/postgres/definitions/tables/spaceship.sql [90m(resource:table:spaceship)[39m[22m

      [1m[33mâ†“[39m[22m [1m[1m[31m[MANUAL_MIGRATION][39m[1m src/contract/__test_assets__/postgres/definitions/tables/spaceship_cargo.sql [90m(resource:table:spaceship_cargo)[39m[22m

      [1m[32mâœ”[39m[22m [1m[1m[33m[REAPPLY][39m[1m src/contract/__test_assets__/postgres/definitions/views/view_spaceship_with_cargo.sql [90m(resource:view:view_spaceship_with_cargo)[39m[22m

"
`;
